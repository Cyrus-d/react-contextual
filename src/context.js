import React from 'react'
import { createContext } from 'react-broadcast'

const providers = new Map()
const Context = createContext()

export function createNamedContext(name, initialState) {
    const context = createContext(initialState)
    providers.set(name, context)
    return context
}

export function getNamedContext(name) {
    return providers.get(name)
}

export function removeNamedContext(name) {
    providers.delete(name)
}

export function resolveContext(context, props) {
    let result
    if (typeof context === 'function') {
        // Test against component-symbol first, then assume a user function
        result = getNamedContext(context) || resolveContext(context(props))
    } else if (typeof context === 'string') {
        result = getNamedContext(context)
    }
    return result || context ||Â Context
}

export default Context