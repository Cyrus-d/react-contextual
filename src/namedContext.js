import React from 'react'
import {
  resolveContext,
  createNamedContext,
  removeNamedContext,
} from './context'

export default function namedContext(contextName, initialState) {
  return Wrapped =>
    class NamedContext extends React.PureComponent {
      constructor(props) {
        super()
        this.name =
          typeof contextName === 'function' ? contextName(props) : contextName
        this.state = { context: createNamedContext(this.name, initialState) }
      }
      componentWillUnmount() {
        removeNamedContext(this.name)
      }
      render() {
        return <Wrapped {...this.props} context={this.state.context} />
      }
    }
}
