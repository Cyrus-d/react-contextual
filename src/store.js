import React from 'react'
import PropTypes from 'prop-types'
import ProviderContext, { createNamedContext, removeNamedContext } from './context'

export class RenderPure extends React.PureComponent {
    render() {
        return this.props.children
    }
}

export class Provider extends React.PureComponent {
    static propTypes = {
        id: PropTypes.string,
        initialState: PropTypes.object.isRequired,
        actions: PropTypes.object,
    }
    constructor(props) {
        super()
        this.state = props.initialState
        this.Context = props.id ? createNamedContext(props.id) : ProviderContext
        if (props.actions) {
            this.actions = Object.keys(props.actions).reduce(
                (acc, name) => ({
                    ...acc,
                    [name]: (...args) =>
                        new Promise(res => {
                            let result = props.actions[name](...args)
                            if (typeof result === 'function')
                                Promise.resolve(result(this.state)).then(state => this.setState(state, res))
                            else this.setState(result, res)
                        }),
                }),
                {},
            )
        }
    }
    componentWillUnmount() {
        if (this.props.id) removeNamedContext(this.props.id)
    }
    render() {
        const { state, actions, props, Context } = this
        const { children, id, actions: _, initialState, ...rest } = props
        const value = { ...state, ...(actions ? { actions } : {}), ...rest }
        return <Context.Provider value={value} children={<RenderPure children={children} />} />
    }
}
