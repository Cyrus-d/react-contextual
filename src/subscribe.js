import React from 'react'
import PropTypes from 'prop-types'
import ProviderContext, { getNamedContext, resolveContext } from './context'

const toArray = obj => Array.isArray(obj) ? obj : [obj]

export function subscribe(...args) {
    return Wrapped =>
        function SubscribeWrap(props) {
            // Filter undefined args (can happen if Subscribe injects them)
            args = args.filter(a => a)
            let contextRefs = ProviderContext,
                mapContextToProps = props => props
            if (args.length === 1) {
                // Check if the argument is a valid context first, if not, assume mapContextToProps
                if (resolveContext(args[0], props, null)) contextRefs = args[0]
                else mapContextToProps = args[0]
            } else if (args.length === 2) {
                // subscribe(Context, mapContextToProps)
                contextRefs = args[0]
                mapContextToProps = args[1]
            }
            if (typeof mapContextToProps !== 'function') {
                // 'theme' or ['theme', 'user', 'language']
                const values = mapContextToProps
                mapContextToProps = (...args) =>
                    toArray(values).reduce((acc, key, index) => ({ ...acc, [key]: args[index] }), {})
            }
            contextRefs = toArray(contextRefs).map(context => resolveContext(context, props))
            let values = []
            return [...contextRefs, Wrapped].reduceRight((accumulator, Context, i) => (
                <Context.Consumer>
                    {value => {
                        values[i] = value
                        if (accumulator === Wrapped) {
                            let context = mapContextToProps(...values, props)
                            context = typeof context === 'object' ? context : { context }
                            values = []
                            return <Wrapped {...props} {...context} />
                        } else return accumulator
                    }}
                </Context.Consumer>
            ))
        }
}

export class Subscribe extends React.PureComponent {
    static propTypes = {
        to: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.object, PropTypes.string, PropTypes.func])),
            PropTypes.object,
            PropTypes.string,
            PropTypes.func,
        ]),
        select: PropTypes.oneOfType([PropTypes.func, PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
        children: PropTypes.func.isRequired,
    }
    static defaultProps = { to: ProviderContext, select: props => props }
    render() {
        const { to, select, children, ...rest } = this.props
        return React.createElement(subscribe(to, select)(props => children(props)), rest)
    }
}
