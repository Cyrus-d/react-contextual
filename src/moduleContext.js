import React from 'react'
import { createNamedContext } from './context'

export default function moduleContext(initialState) {
    return Wrapped => {
        let context = undefined
        const ModuleContext = class extends React.PureComponent {
            render() {
                return <Wrapped {...this.props} context={context} />
            }
        }
        context = createNamedContext(ModuleContext, initialState)
        return ModuleContext
    }
}