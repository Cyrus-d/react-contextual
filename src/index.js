import { subscribe, Subscribe } from './subscribe'
import { RenderPure, Provider } from './store'
import Context, {
    namedContext,
    moduleContext,
    transformContext,
    createNamedContext,
    removeNamedContext,
    getNamedContext,
    resolveContext,
} from './context'

export {
    Context,
    subscribe,
    Subscribe,
    RenderPure,
    Provider,
    namedContext,
    moduleContext,
    transformContext,
    createNamedContext,
    removeNamedContext,
    getNamedContext,
}
