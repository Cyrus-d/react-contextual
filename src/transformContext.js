import React from 'react'
import { subscribe } from './subscribe'
import { resolveContext } from './context'

export default function transformContext(context, transform) {
    return Wrapped => {
        return subscribe(context, transform)(class extends React.PureComponent {
            render() {
                const context = resolveContext(context, this.props)
                return <Wrapped {...this.props} context={context} />
            }
        })
    }
}