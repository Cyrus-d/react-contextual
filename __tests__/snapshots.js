import 'raf/polyfill'
import React from 'react'
import Adapter from 'enzyme-adapter-react-16'
import Enzyme, { shallow, mount } from 'enzyme'
import delay from 'delay'
import { Provider, Context, Subscribe, subscribe } from '../src/'

Enzyme.configure({ adapter: new Adapter() })

const store = {
    initialState: { count: 0 },
    actions: { up: () => state => ({ count: state.count + 1 }) },
}

async function snapshot(Root, mutation) {
    const tree = mount(Root)
    expect(tree).toMatchSnapshot()
    if (mutation) {
        mutation(tree)
        await delay(30)
        tree.update()
        expect(tree).toMatchSnapshot()
    }
}

test('subscribe()', async () => {
    const Test = subscribe()(props => <button onClick={props.actions.up}>{props.count}</button>)
    await snapshot(
        <Provider {...store}>
            <Test />
            <Subscribe children={props => props.count} />
        </Provider>,
        tree => tree.find('button').simulate('click'),
    )
})

test('subscribe(props => props)', async () => {
    const Test = subscribe(props => props)(props => <button onClick={props.actions.up}>{props.count}</button>)
    await snapshot(
        <Provider {...store}>
            <Test />
            <Subscribe select={props => props} children={props => props.count} />
        </Provider>,
        tree => tree.find('button').simulate('click'),
    )
})

test('subscribe(id, props => props)', async () => {
    const Test = subscribe('id', props => props)(props => <button onClick={props.actions.up}>{props.count}</button>)
    await snapshot(
        <Provider id="id" {...store}>
            <Test />
            <Subscribe to="id" select={props => props} children={props => props.count} />
        </Provider>,
        tree => tree.find('button').simulate('click'),
    )
})

test('subscribe(props => props.id, props => props)', async () => {
    const Test = subscribe(props => props.id, props => props)(props => (
        <button onClick={props.actions.up}>{props.count}</button>
    ))
    await snapshot(
        <Provider id="id" {...store}>
            <Test id="id" />
            <Subscribe to="id" select={props => props} children={props => props.count} />
        </Provider>,
        tree => tree.find('button').simulate('click'),
    )
})

test('subscribe(Context, props => props)', async () => {
    const Test = subscribe(Context, props => props)(props => <button onClick={props.actions.up}>{props.count}</button>)
    await snapshot(
        <Provider {...store}>
            <Test />
            <Subscribe to={Context} select={props => props} children={props => props.count} />
        </Provider>,
        tree => tree.find('button').simulate('click'),
    )
})

test('subscribe([a,b], (a,b) => props)', async () => {
    const Test = subscribe(['id1', 'id2'], (store1, store2) => ({ store1, store2 }))(props => (
        <button onClick={() => props.store1.actions.up() && props.store2.actions.up()}>
            {props.store1.count} {props.store2.count}
        </button>
    ))
    await snapshot(
        <Provider id="id1" {...store}>
            <Provider id="id2" {...store}>
                <Test />
                <Subscribe
                    to={['id1', 'id2']}
                    select={(store1, store2) => ({ store1, store2 })}
                    children={props => props.store1.count + " " + props.store2.count}
                />
            </Provider>
        </Provider>,
        tree => tree.find('button').simulate('click'),
    )
})
