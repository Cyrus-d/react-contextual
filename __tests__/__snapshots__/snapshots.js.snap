// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subscribe() 1`] = `
<Provider
  actions={
    Object {
      "up": [Function],
    }
  }
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "up": [Function],
        },
        "count": 0,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            actions={
              Object {
                "up": [Function],
              }
            }
            count={0}
          >
            <button
              onClick={[Function]}
            >
              0
            </button>
          </Component>
        </Consumer>
      </Component>
      <Subscribe
        select={[Function]}
        to={
          Object {
            "Consumer": [Function],
            "Provider": [Function],
          }
        }
      >
        <Component>
          <Consumer
            quiet={false}
          >
            <Component
              actions={
                Object {
                  "up": [Function],
                }
              }
              count={0}
            >
              0
            </Component>
          </Consumer>
        </Component>
      </Subscribe>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`subscribe() 2`] = `
<Provider
  actions={
    Object {
      "up": [Function],
    }
  }
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "up": [Function],
        },
        "count": 1,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            actions={
              Object {
                "up": [Function],
              }
            }
            count={1}
          >
            <button
              onClick={[Function]}
            >
              1
            </button>
          </Component>
        </Consumer>
      </Component>
      <Subscribe
        select={[Function]}
        to={
          Object {
            "Consumer": [Function],
            "Provider": [Function],
          }
        }
      >
        <Component>
          <Consumer
            quiet={false}
          >
            <Component
              actions={
                Object {
                  "up": [Function],
                }
              }
              count={1}
            >
              1
            </Component>
          </Consumer>
        </Component>
      </Subscribe>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`subscribe([a,b], (a,b) => props) 1`] = `
<Provider
  actions={
    Object {
      "up": [Function],
    }
  }
  id="id1"
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "up": [Function],
        },
        "count": 0,
      }
    }
  >
    <RenderPure>
      <Provider
        actions={
          Object {
            "up": [Function],
          }
        }
        id="id2"
        initialState={
          Object {
            "count": 0,
          }
        }
      >
        <Provider
          value={
            Object {
              "actions": Object {
                "up": [Function],
              },
              "count": 0,
            }
          }
        >
          <RenderPure>
            <Component>
              <Consumer
                quiet={false}
              >
                <Consumer
                  quiet={false}
                >
                  <Component
                    store1={
                      Object {
                        "actions": Object {
                          "up": [Function],
                        },
                        "count": 0,
                      }
                    }
                    store2={
                      Object {
                        "actions": Object {
                          "up": [Function],
                        },
                        "count": 0,
                      }
                    }
                  >
                    <button
                      onClick={[Function]}
                    >
                      0
                       
                      0
                    </button>
                  </Component>
                </Consumer>
              </Consumer>
            </Component>
            <Subscribe
              select={[Function]}
              to={
                Array [
                  "id1",
                  "id2",
                ]
              }
            >
              <Component>
                <Consumer
                  quiet={false}
                >
                  <Consumer
                    quiet={false}
                  >
                    <Component
                      store1={
                        Object {
                          "actions": Object {
                            "up": [Function],
                          },
                          "count": 0,
                        }
                      }
                      store2={
                        Object {
                          "actions": Object {
                            "up": [Function],
                          },
                          "count": 0,
                        }
                      }
                    >
                      0 0
                    </Component>
                  </Consumer>
                </Consumer>
              </Component>
            </Subscribe>
          </RenderPure>
        </Provider>
      </Provider>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`subscribe([a,b], (a,b) => props) 2`] = `
<Provider
  actions={
    Object {
      "up": [Function],
    }
  }
  id="id1"
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "up": [Function],
        },
        "count": 1,
      }
    }
  >
    <RenderPure>
      <Provider
        actions={
          Object {
            "up": [Function],
          }
        }
        id="id2"
        initialState={
          Object {
            "count": 0,
          }
        }
      >
        <Provider
          value={
            Object {
              "actions": Object {
                "up": [Function],
              },
              "count": 1,
            }
          }
        >
          <RenderPure>
            <Component>
              <Consumer
                quiet={false}
              >
                <Consumer
                  quiet={false}
                >
                  <Component
                    store1={
                      Object {
                        "actions": Object {
                          "up": [Function],
                        },
                        "count": 1,
                      }
                    }
                    store2={
                      Object {
                        "actions": Object {
                          "up": [Function],
                        },
                        "count": 1,
                      }
                    }
                  >
                    <button
                      onClick={[Function]}
                    >
                      1
                       
                      1
                    </button>
                  </Component>
                </Consumer>
              </Consumer>
            </Component>
            <Subscribe
              select={[Function]}
              to={
                Array [
                  "id1",
                  "id2",
                ]
              }
            >
              <Component>
                <Consumer
                  quiet={false}
                >
                  <Consumer
                    quiet={false}
                  >
                    <Component
                      store1={
                        Object {
                          "actions": Object {
                            "up": [Function],
                          },
                          "count": 1,
                        }
                      }
                      store2={
                        Object {
                          "actions": Object {
                            "up": [Function],
                          },
                          "count": 1,
                        }
                      }
                    >
                      1 1
                    </Component>
                  </Consumer>
                </Consumer>
              </Component>
            </Subscribe>
          </RenderPure>
        </Provider>
      </Provider>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`subscribe(Context, props => props) 1`] = `
<Provider
  actions={
    Object {
      "up": [Function],
    }
  }
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "up": [Function],
        },
        "count": 0,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            actions={
              Object {
                "up": [Function],
              }
            }
            count={0}
          >
            <button
              onClick={[Function]}
            >
              0
            </button>
          </Component>
        </Consumer>
      </Component>
      <Subscribe
        select={[Function]}
        to={
          Object {
            "Consumer": [Function],
            "Provider": [Function],
          }
        }
      >
        <Component>
          <Consumer
            quiet={false}
          >
            <Component
              actions={
                Object {
                  "up": [Function],
                }
              }
              count={0}
            >
              0
            </Component>
          </Consumer>
        </Component>
      </Subscribe>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`subscribe(Context, props => props) 2`] = `
<Provider
  actions={
    Object {
      "up": [Function],
    }
  }
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "up": [Function],
        },
        "count": 1,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            actions={
              Object {
                "up": [Function],
              }
            }
            count={1}
          >
            <button
              onClick={[Function]}
            >
              1
            </button>
          </Component>
        </Consumer>
      </Component>
      <Subscribe
        select={[Function]}
        to={
          Object {
            "Consumer": [Function],
            "Provider": [Function],
          }
        }
      >
        <Component>
          <Consumer
            quiet={false}
          >
            <Component
              actions={
                Object {
                  "up": [Function],
                }
              }
              count={1}
            >
              1
            </Component>
          </Consumer>
        </Component>
      </Subscribe>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`subscribe(id, props => props) 1`] = `
<Provider
  actions={
    Object {
      "up": [Function],
    }
  }
  id="id"
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "up": [Function],
        },
        "count": 0,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            actions={
              Object {
                "up": [Function],
              }
            }
            count={0}
          >
            <button
              onClick={[Function]}
            >
              0
            </button>
          </Component>
        </Consumer>
      </Component>
      <Subscribe
        select={[Function]}
        to="id"
      >
        <Component>
          <Consumer
            quiet={false}
          >
            <Component
              actions={
                Object {
                  "up": [Function],
                }
              }
              count={0}
            >
              0
            </Component>
          </Consumer>
        </Component>
      </Subscribe>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`subscribe(id, props => props) 2`] = `
<Provider
  actions={
    Object {
      "up": [Function],
    }
  }
  id="id"
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "up": [Function],
        },
        "count": 1,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            actions={
              Object {
                "up": [Function],
              }
            }
            count={1}
          >
            <button
              onClick={[Function]}
            >
              1
            </button>
          </Component>
        </Consumer>
      </Component>
      <Subscribe
        select={[Function]}
        to="id"
      >
        <Component>
          <Consumer
            quiet={false}
          >
            <Component
              actions={
                Object {
                  "up": [Function],
                }
              }
              count={1}
            >
              1
            </Component>
          </Consumer>
        </Component>
      </Subscribe>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`subscribe(props => props) 1`] = `
<Provider
  actions={
    Object {
      "up": [Function],
    }
  }
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "up": [Function],
        },
        "count": 0,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            actions={
              Object {
                "up": [Function],
              }
            }
            count={0}
          >
            <button
              onClick={[Function]}
            >
              0
            </button>
          </Component>
        </Consumer>
      </Component>
      <Subscribe
        select={[Function]}
        to={
          Object {
            "Consumer": [Function],
            "Provider": [Function],
          }
        }
      >
        <Component>
          <Consumer
            quiet={false}
          >
            <Component
              actions={
                Object {
                  "up": [Function],
                }
              }
              count={0}
            >
              0
            </Component>
          </Consumer>
        </Component>
      </Subscribe>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`subscribe(props => props) 2`] = `
<Provider
  actions={
    Object {
      "up": [Function],
    }
  }
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "up": [Function],
        },
        "count": 1,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            actions={
              Object {
                "up": [Function],
              }
            }
            count={1}
          >
            <button
              onClick={[Function]}
            >
              1
            </button>
          </Component>
        </Consumer>
      </Component>
      <Subscribe
        select={[Function]}
        to={
          Object {
            "Consumer": [Function],
            "Provider": [Function],
          }
        }
      >
        <Component>
          <Consumer
            quiet={false}
          >
            <Component
              actions={
                Object {
                  "up": [Function],
                }
              }
              count={1}
            >
              1
            </Component>
          </Consumer>
        </Component>
      </Subscribe>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`subscribe(props => props.id, props => props) 1`] = `
<Provider
  actions={
    Object {
      "up": [Function],
    }
  }
  id="id"
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "up": [Function],
        },
        "count": 0,
      }
    }
  >
    <RenderPure>
      <Component
        id="id"
      >
        <Consumer
          quiet={false}
        >
          <Component
            actions={
              Object {
                "up": [Function],
              }
            }
            count={0}
            id="id"
          >
            <button
              onClick={[Function]}
            >
              0
            </button>
          </Component>
        </Consumer>
      </Component>
      <Subscribe
        select={[Function]}
        to="id"
      >
        <Component>
          <Consumer
            quiet={false}
          >
            <Component
              actions={
                Object {
                  "up": [Function],
                }
              }
              count={0}
            >
              0
            </Component>
          </Consumer>
        </Component>
      </Subscribe>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`subscribe(props => props.id, props => props) 2`] = `
<Provider
  actions={
    Object {
      "up": [Function],
    }
  }
  id="id"
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "up": [Function],
        },
        "count": 1,
      }
    }
  >
    <RenderPure>
      <Component
        id="id"
      >
        <Consumer
          quiet={false}
        >
          <Component
            actions={
              Object {
                "up": [Function],
              }
            }
            count={1}
            id="id"
          >
            <button
              onClick={[Function]}
            >
              1
            </button>
          </Component>
        </Consumer>
      </Component>
      <Subscribe
        select={[Function]}
        to="id"
      >
        <Component>
          <Consumer
            quiet={false}
          >
            <Component
              actions={
                Object {
                  "up": [Function],
                }
              }
              count={1}
            >
              1
            </Component>
          </Consumer>
        </Component>
      </Subscribe>
    </RenderPure>
  </Provider>
</Provider>
`;
